src/main/java/org/breedinginsight/brapps/importer/services/processors/ExperimentProcessor.java

```diff
@@ -0,0 +1,708 @@
+            String key = createObservationUnitKey(importRow);
+            this.observationUnitByNameNoScope.put(key, obsUnitPIO);
+        }
+    }
+
+    private String createObservationUnitKey(ExperimentObservation importRow) {
+        String key = importRow.getEnv() + importRow.getExpUnitId();
+        return key;
+    }
+
+    private ValidationErrors validateFields(List<BrAPIImport> importRows, ValidationErrors validationErrors) {
+        HashSet<String> uniqueStudyAndObsUnit = new HashSet<>();
+        for (int i = 0; i < importRows.size(); i++) {
+            ExperimentObservation importRow = (ExperimentObservation) importRows.get(i);
+            validateConditionallyRequired(validationErrors, i, importRow);
+            validateUniqueObsUnits(validationErrors, uniqueStudyAndObsUnit, i, importRow);
+        }
+        return validationErrors;
+    }
+
+    /**
+     * Validate that the the observation unit is unique within a study.
+     * <br>
+     * SIDE EFFECTS:  validationErrors and uniqueStudyAndObsUnit can be modified.
+     * @param validationErrors can be modified as a side effect.
+     * @param uniqueStudyAndObsUnit can be modified as a side effect.
+     * @param i counter that is always two less the file row being validated
+     * @param importRow the data row being validated
+     */
+    private void validateUniqueObsUnits(
+            ValidationErrors validationErrors,
+            HashSet<String> uniqueStudyAndObsUnit,
+            int i,
+            ExperimentObservation importRow) {
+        String envIdPlusStudyId = createObservationUnitKey( importRow );
+        if( uniqueStudyAndObsUnit.contains( envIdPlusStudyId )){
+            String errorMessage = String.format("The ID (%s) is not unique within the environment(%s)", importRow.getExpUnitId(), importRow.getEnv());
+            this.addRowError("Exp Unit ID", errorMessage, validationErrors, i);
+        }
+        else{
+            uniqueStudyAndObsUnit.add( envIdPlusStudyId );
+        }
+    }
+
+    private void validateConditionallyRequired(ValidationErrors validationErrors, int i, ExperimentObservation importRow) {
+        String experimentTitle = importRow.getExpTitle();
+        String obsUnitID = importRow.getObsUnitID();
+        if( StringUtils.isBlank( obsUnitID )){
+            validateRequiredCell(
+                    experimentTitle,
+                    "Exp Title",
+                    "Field is blank", validationErrors, i
+                    );
+        }
+
+        ImportObjectState expState = this.trialByNameNoScope.get(experimentTitle).getState();
+        boolean isExperimentNew = (expState == ImportObjectState.NEW);
+
+        if (isExperimentNew) {
+            String errorMessage = "Field is blank when creating a new experiment";
+            validateRequiredCell( importRow.getGid(),
+                    "GID",
+                    errorMessage, validationErrors, i);
```

Comment:
I believe if you get to this error you will have 2 GID errors. One for your missing GID check in the `process` method and then one here.

Reply:
the process() method guards against this with the logic:

>            if (! StringUtils.isBlank( importRow.getGid() )) { // if GID is blank, don't bother to check if it is valid.
>                 validateGermplasm(importRow,validationErrors, i, germplasmPIO);
>             }

Reply:
Sounds good!